### ISSUE TITLE
False Positive: Trivial `super.equals` delegation and `Comparable` special cases are treated as real pairing violations.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5632]
# False Positive: Trivial `super.equals` delegation and `Comparable` special cases are treated as real pairing violations.

Version
2.48.0

## Checker
- Checker id: `EqualsHashCode`
- Checker description: Detects classes that override equals(Object) in a non-trivial way but do not override hashCode().

## Description of the false positive
`diff/output` marks 3 false positives for `EqualsHashCode`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The recurring themes are Comparable-specific handling, abstract declarations, suppression or annotations.

### `NegCase3.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/neg/NegCase3.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. Reporting this as if the class had already committed to a concrete contract is too aggressive. That matches the benchmark's stated intent: A class with abstract equals method should not be flagged as missing hashCode.

```java
// A class with abstract equals method should not be flagged as missing hashCode.
package scensct.core.neg;

public abstract class NegCase3 {
    public abstract boolean equals(Object obj);
}
```

### `NegCase2.java`
Source: `output/pmd/OverrideBothEqualsAndHashcode/src/main/java/scensct/core/neg/NegCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The benchmark is using `Comparable` as an exclusion signal here, so a report reads as policy mismatch rather than a strong correctness issue. That matches the benchmark's stated intent: A non-Comparable class contains a bridge method named "equals" with a single parameter of type Object should not be flagged as overriding equals without hashCode.

```java
// A non-Comparable class contains a bridge method named "equals" with a single parameter of type Object should not be flagged as overriding equals without hashCode.
package scensct.core.neg;

public class NegCase2 implements Comparable<NegCase2> {
    // Bridge method generated by compiler for generic scenarios.
    // This is a synthetic/bridge method; the checker should ignore it.
    public boolean equals(Object obj) { // [REPORTED LINE]
        return false;
    }
    // The presence of this bridge method alone should not trigger the checker.
    public int compareTo(NegCase2 other) {
        return 0;
    }
}
```

### `NegCase7.java`
Source: `output/pmd/OverrideBothEqualsAndHashcode/src/main/java/scensct/core/neg/NegCase7.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The benchmark is using `Comparable` as an exclusion signal here, so a report reads as policy mismatch rather than a strong correctness issue. That matches the benchmark's stated intent: A class that implements Comparable should not be flagged as overriding equals without hashCode.

```java
// A class that implements Comparable should not be flagged as overriding equals without hashCode.
package scensct.core.neg;

public class NegCase7 implements Comparable<NegCase7> {
    // Overrides equals without hashCode, but class implements Comparable.
    // Checker should allow this.
    @Override
    public boolean equals(Object obj) {
        return false;
    }
    
    @Override
    public int compareTo(NegCase7 other) {
        return 0;
    }
}
```

## Cause analysis
The rule is documented as follows: Detects classes that override equals(Object) in a non-trivial way but do not override hashCode().

In this set the problem shows up around Comparable-specific handling, abstract declarations, suppression or annotations. The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `EqualsHashCode`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
